# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieParser.cpp
# Commit: b668d0c685e8
# Full Hash: b668d0c685e8d0bbcd3bf5b1856389ca41be31f3
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2025-01-09 04:16:57
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r 08acd8256071 -r b668d0c685e8 netwerk/cookie/CookieParser.cpp
--- a/netwerk/cookie/CookieParser.cpp	Wed Jan 08 19:25:59 2025 +0000
+++ b/netwerk/cookie/CookieParser.cpp	Wed Jan 08 19:25:59 2025 +0000
@@ -858,7 +858,7 @@
                          CookieStatus aStatus, nsCString& aCookieHeader,
                          const nsACString& aDateHeader, bool aFromHttp,
                          bool aIsForeignAndNotAddon, bool aPartitionedOnly,
-                         bool aIsInPrivateBrowsing) {
+                         bool aIsInPrivateBrowsing, bool aOn3pcbException) {
   MOZ_ASSERT(!mContainsCookie);
 
   // init expiryTime such that session cookies won't prematurely expire
@@ -1005,6 +1005,19 @@
     return;
   }
 
+  // If the cookie is on the 3pcd exception list, we apply partitioned
+  // attribute to the cookie.
+  if (aOn3pcbException) {
+    // We send a warning if the cookie doesn't have the partitioned attribute
+    // in the foreign context.
+    if (aPartitionedOnly && !mCookieData.isPartitioned() &&
+        aIsForeignAndNotAddon) {
+      mWarnings.mForeignNoPartitionedWarning = true;
+    }
+
+    mCookieData.isPartitioned() = true;
+  }
+
   // If the cookie does not have the partitioned attribute,
   // but is foreign we should give the developer a message.
   // If CHIPS isn't required yet, we will warn the console
@@ -1020,6 +1033,7 @@
       RejectCookie(RejectedForeignNoPartitionedError);
       return;
     }
+
     mWarnings.mForeignNoPartitionedWarning = true;
   }
 